### DISCUSSION TITLE
Convergence issue

### DISCUSSION THREAD

**Discussion by woojinjung-opt** [idaholab/moose#27034] [Q&A General]
I'm making phase-field based app using KKS. However, the solution does not converge. Also, there is error related with ConstantDT. Could you give me some help? This is the output message :

Time Step 1, time = 1.49012e-12, dt = 1.49012e-12
 0 Nonlinear |R| = 1.920919e-05
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_NUMERIC_ZEROPIVOT 
Nonlinear solve did not converge due to DIVERGED_LOCAL_MIN iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge

Time Step 1, time = 1e-12, dt = 1e-12
 0 Nonlinear |R| = 1.920919e-05
  Linear solve did not converge due to DIVERGED_PC_FAILED iterations 0
                 PC failed due to FACTOR_NUMERIC_ZEROPIVOT 
Nonlinear solve did not converge due to DIVERGED_LOCAL_MIN iterations 0
 Solve Did NOT Converge!
Aborting as solve did not converge

*** ERROR ***
The following error occurred in the object "ConstantDT", of type "ConstantDT".

Solve failed and timestep already at or below dtmin, cannot continue!

I'll attach full input file.
[redistribution.txt](https://github.com/idaholab/moose/files/14536204/redistribution.txt)

**Comment by GiudGiud** [comment]
Hello

Please follow the advice on the troubleshooting failed page on the moose website and report on what you tried

Guillaume

**Reply by Leeyouan** [reply]
Sorry for late.

I tried some method. There is issue on initial condition & unit system.
Actually, I have another question in Cahn-Hilliard equation.

What I want is that adapting temperature gradient effect on Cahn-Hilliard equation. 
This is the equation.

![Selection_017](https://github.com/idaholab/moose/assets/125252329/bece8c69-e10b-44ab-b136-a366122596e1)

How can I add temperature gradient effect in MOOSE?

Thank you.

Woojin

**Reply by GiudGiud** [reply]
Hello

You'll want to create a new kernel if there is none implemented in the phase field module.
This does not seem too complex to implement but it would be worth looking through examples. Search for a "T" or "temperature" variable defined in CH inputs

Guillaume

**Reply by Leeyouan** [reply]
Thank you for replying.

I'm not making kernel for that. But I'm just wondering that is it possible to adapt like this?
```
  [./CHBulk]
    type = KKSSplitCHCRes
    variable = c
    ca       = cab
    fa_name  = fab
    w        = w
  [../]

  [./SoretDiffusion]
    type = SoretDiffusion
    variable = w
    c = c
    T = T
    diff_name = M_t
    Q_name = Qstar
  [../]

  [./dcdt]
    type = CoupledTimeDerivative
    variable = w
    v = c
  [../]
  [./ckernel]
    type = SplitCHWRes
    mob_name = M
    variable = w
  [../]

  [./HeatConduction]
    type = MatDiffusion
    variable = T
    diffusivity = k
  [../]
```
I'm really appreciate to your answer.

Woojin

**Reply by GiudGiud** [reply]
Hello

I dont see the M_c term? Maybe use comments in the inputs to indicate which term is which

Guillaume

**Reply by Leeyouan** [reply]
Oh. The term 'M' in the 'SplitCHWRes' denotes M_c. I must revise it to M_c.

The solution converges well nonetheless reliability of the result. I'm now going to check it.

Thank you for help.

Woojin

**Reply by Leeyouan** [reply]
I found it does not work well..!

If I set temperature profile in first order, that should be no change in the result cause it has two gradient.

But it represents different result even though I set first order temperature profile.

I should make new kernel. Thanks.

Woojin
